instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java

```diff
@@ -39,52 +38,41 @@ public static void onEnter(
-          helper().updateContext(attachedContext, httpServletRequest, mappingResolver, servlet);
-      scope = attachedContext.makeCurrent();
-      // We are inside nested servlet/filter/app-server span, don't create new span
-      return;
-    }
-
-    if (attachedContext != null || ServerSpan.fromContextOrNull(currentContext) != null) {
-      // Update context with info from current request to ensure that server span gets the best
-      // possible name.
-      // In case server span was created by app server instrumentations calling updateContext
-      // returns a new context that contains servlet context path that is used in other
-      // instrumentations for naming server span.
-      MappingResolver mappingResolver = Servlet3Singletons.getMappingResolver(servletOrFilter);
-      boolean servlet = servletOrFilter instanceof Servlet;
-      Context updatedContext =
-          helper().updateContext(currentContext, httpServletRequest, mappingResolver, servlet);
-      if (currentContext != updatedContext) {
-        // updateContext updated context, need to re-scope
-        scope = updatedContext.makeCurrent();
-      }
-      // We are inside nested servlet/filter/app-server span, don't create new span
-      return;
-    }
+    Context contextToUpdate;
 
     requestContext = new ServletRequestContext<>(httpServletRequest, servletOrFilter);
-
-    if (!helper().shouldStart(currentContext, requestContext)) {
-      return;
+    if (attachedContext == null && helper().shouldStart(currentContext, requestContext)) {
+      context = helper().start(currentContext, requestContext);
+      helper().setAsyncListenerResponse(httpServletRequest, (HttpServletResponse) response);
+
+      contextToUpdate = context;
+    } else if (attachedContext != null
+        && helper().needsRescoping(currentContext, attachedContext)) {
+      // Given request already has a context associated with it.
+      // see the needsRescoping() javadoc for more explanation
+      contextToUpdate = attachedContext;
+    } else {
+      // We are inside nested servlet/filter/app-server span, don't create new span
+      contextToUpdate = currentContext;
     }
 
-    context = helper().start(currentContext, requestContext);
-    scope = context.makeCurrent();
-
-    helper().setAsyncListenerResponse(httpServletRequest, (HttpServletResponse) response);
+    // Update context with info from current request to ensure that server span gets the best
+    // possible name.
+    // In case server span was created by app server instrumentations calling updateContext
+    // returns a new context that contains servlet context path that is used in other
+    // instrumentations for naming server span.
+    MappingResolver mappingResolver = Servlet3Singletons.getMappingResolver(servletOrFilter);
+    boolean servlet = servletOrFilter instanceof Servlet;
+    contextToUpdate =
+        helper().updateContext(contextToUpdate, httpServletRequest, mappingResolver, servlet);
```

Comment:
> the previous behavior was nice of setting everything up correctly in `startSpan` (e.g. correct span name and correct `ServerSpanNaming.Source`)

Honestly I found setting the `Source` a bit confusing: what if your span name was `HTTP GET` with `FILTER` as a source? If you had another filter trying to update the name to `/abc` it wouldn't, because it was a shorter name - which I'm not so sure is correct; setting the source seems to be closely connected to the route, not necessarily the span name (even if they're the same in most cases). (Although I'm not sure how "real" this example is 😅 )

Also, the only 2 instrumentations I found that have access to the route in the beginning are armeria and servlets; and servlet span name will usually be updated by something like spring or jax-rs. Which implies that for most cases, the route should be extracted on `end()` (and span name updated somewhere in the meantime).